providers/azure_oidc.go

```diff
@@ -0,0 +1,201 @@
+package providers
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/pusher/oauth2_proxy/pkg/apis/sessions"
+	"github.com/pusher/oauth2_proxy/pkg/logger"
+	"github.com/pusher/oauth2_proxy/pkg/requests"
+	"golang.org/x/oauth2"
+)
+
+type AzureOIDCProvider struct {
+	*OIDCProvider
+	PermittedGroups []string
+	ExemptedEmails  []string
+}
+
+type AzureOIDCClaims struct {
+	*OIDCClaims
+	Groups            []string `json:"groups"`
+	PreferredUsername string   `json:"preferred_username"`
+}
+
+// NewAzureOIDCProvider initiates a new AzureOIDCProvider
+func NewAzureOIDCProvider(p *ProviderData) *AzureOIDCProvider {
+	provider := AzureOIDCProvider{
+		OIDCProvider: NewOIDCProvider(p),
+	}
+	provider.ProviderName = "Azure OpenID Connect"
+	if p.Scope == "" {
+		provider.Scope = "openid email profile"
+	}
+
+	if provider.ProfileURL == nil || provider.ProfileURL.String() == "" {
+		provider.ProfileURL = &url.URL{
+			Scheme: "https",
+			Host:   "graph.microsoft.com",
+			Path:   "/v1.0/me",
```

Comment:
There's a comment in the docs saying that the `v2.0` API should be used, but the default is `v1.0` by the looks of this, any particular reason why?

Reply:
v2.0 refers to OIDC endpoint `https://login.microsoftonline.com/<<aplication_id>>/v2.0`

Reply:
I checked out the OIDC discovery end point for this based on the v2 URL in the documentation on this PR. The discovery URL suggests the userinfo endpoint is `https://graph.microsoft.com/oidc/userinfo` so perhaps this it the v2 API version of the `v1.0/me endpoint`? WDYT?

Reply:
Makes sense